gui/widgets/grid.cpp

```diff
@@ -0,0 +1,640 @@
+		}
+	}
+	return nullptr;
+}
+
+const Graphics::ManagedSurface *GridWidget::languageToSurface(Common::Language languageCode) {
+	if (languageCode == Common::Language::UNK_LANG)
+		return nullptr;
+	return _languageIcons[languageCode];
+}
+
+const Graphics::ManagedSurface *GridWidget::platformToSurface(Common::Platform platformCode) {
+	if (platformCode == Common::Platform::kPlatformUnknown)
+		return nullptr;
+	return _platformIcons[platformCode];
+}
+
+void GridWidget::setEntryList(Common::Array<GridItemInfo> *list) {
+	_allEntries.clear();
+	for (Common::Array<GridItemInfo>::iterator entryIter = list->begin(); entryIter != list->end(); ++entryIter) {
+		_allEntries.push_back(*entryIter);
+	}
+	if (!_gridItems.empty()) {
+		reflowLayout();
+	}
+}
+
+bool GridWidget::calcVisibleEntries() {
+	bool needsReload = false;
+
+	int nFirstVisibleItem = 0, nItemsOnScreen = 0;
+	
+	nFirstVisibleItem = _itemsPerRow * (-_scrollPos / (_gridItemHeight + _gridYSpacing));
+	nFirstVisibleItem = (nFirstVisibleItem < 0) ? 0 : nFirstVisibleItem;
+	
+	nItemsOnScreen = (3 + (_scrollWindowHeight / (_gridItemHeight + _gridYSpacing))) * (_itemsPerRow);
+
+	if (nFirstVisibleItem != _firstVisibleItem || nItemsOnScreen != _itemsOnScreen) {
+		needsReload = true;
+		_itemsOnScreen = nItemsOnScreen;
+		_firstVisibleItem = nFirstVisibleItem;
+
+		int toRender = MIN(_firstVisibleItem + _itemsOnScreen, (int)_allEntries.size());
+
+		_visibleEntries.clear();
+		for (int ind = _firstVisibleItem; ind < toRender; ++ind) {
+			GridItemInfo *iter = _allEntries.begin() + ind;
+			_visibleEntries.push_back(*iter);
+		}
+	}
+	return needsReload;
+}
+
+void GridWidget::setTitlesVisible(bool vis) {
+	_isTitlesVisible = vis;
+}
+
+void GridWidget::reloadThumbnails() {
+	Graphics::ManagedSurface *surf = nullptr;
+	String gameid;
+	String engineid;
+	
+	for (Common::Array<GridItemInfo>::iterator iter = _visibleEntries.begin(); iter != _visibleEntries.end(); ++iter) {
+		if (!_loadedSurfaces.contains(iter->thumbPath)) {
+			surf = loadSurfaceFromFile(iter->thumbPath);
+			if (surf) {
+				const Graphics::ManagedSurface *scSurf(scaleGfx(surf, _thumbnailWidth, 512));
+				_loadedSurfaces[iter->thumbPath] = scSurf;
+				surf->free();
+				delete surf;
+			} else {
+				_loadedSurfaces[iter->thumbPath] = nullptr;
+			}
+		}
+	}
+}
+
+void GridWidget::loadFlagIcons() {
+	const Common::LanguageDescription *l = Common::g_languages;
+	for (; l->code; ++l) {
+		String path = String::format("flags/%s.svg", l->code);
+		Graphics::ManagedSurface *gfx = loadSurfaceFromFile(path);
+		if (gfx) {
+			const Graphics::ManagedSurface *scGfx = scaleGfx(gfx, 32, 128);
+			_languageIcons[l->id] = scGfx;
+			gfx->free();
+			delete gfx;
+		} else {
+			_languageIcons[l->id] = nullptr;
+		}
+	}
+}
+
+void GridWidget::loadPlatformIcons() {
+	const Common::PlatformDescription *l = Common::g_platforms;
+	for (; l->code; ++l) {
+		String path = String::format("platforms/%s.png", l->code);
```

Comment:
This logic seems largely duplicated with loadFlagIcons, consider outfactoring a helper-function.